src/hello_world.cpp

```diff
@@ -26,17 +32,161 @@ NAN_METHOD(HelloWorld::New)
+ * });
+ *
+ */
+
+// this is the cpp object that will be passed around in our method and callbacks
+// referred to as a "baton"
+class AsyncBaton
 {
-    static Nan::Persistent<v8::Function> init;
-    return init;
+  public:
+    uv_work_t request; // required
+    Nan::Persistent<v8::Function> cb; // callback function type
+    std::string phrase;
+    bool louder;
+    std::string error_name;
+    std::string result;
+};
+
+NAN_METHOD(HelloWorld::shout)
+{
+    std::string phrase = "";
+    bool louder = false;
+
+    // check first argument, should be a 'phrase' string
+    if (!info[0]->IsString()) 
+    {
+        Nan::ThrowTypeError("first arg 'phrase' must be a string");
+        return;
+    }
+    phrase = *v8::String::Utf8Value((info[0])->ToString());
+
+    // check second argument, should be an 'options' object
+    if (!info[1]->IsObject()) 
+    {
+        Nan::ThrowTypeError("second arg 'options' must be an object");
+        return;
+    }
+
+    v8::Local<v8::Object> options = info[1].As<v8::Object>();
+    if (options->Has(Nan::New("louder").ToLocalChecked())) 
+    {
+        v8::Local<v8::Value> louder_val = options->Get(Nan::New("louder").ToLocalChecked());
+        if (!louder_val->IsBoolean())
+        {
+            Nan::ThrowError("option 'louder' must be a boolean");
+            return;
+        }
+        louder = louder_val->BooleanValue();
+    }
+
+    // check third argument, should be a 'callback' function
+    if (!info[2]->IsFunction()) 
+    {
+      Nan::ThrowTypeError("third arg 'callback' must be a function");
+      return;
+    }
+    v8::Local<v8::Value> callback = info[2];
+
+    // set up the baton to pass into our threadpool
+    AsyncBaton *baton = new AsyncBaton();
+    baton->request.data = baton;
+    baton->phrase = phrase;
+    baton->louder = louder;
+    baton->cb.Reset(callback.As<v8::Function>());
+
+    // this is the all-important way to pass info into the threadpool using uv_queue_work
+    // it cannot take v8 objects, so we need to do some manipulation above to convert into cpp objects
+    // otherwise things get janky
+    // takes four arguments:
+    // 1) which loop to use, node only has one so we pass in a pointer to the default
+    // 2) the baton defined above, we use this to access information important for the method
+    // 3) operations to be executed within the threadpool
+    // 4) operations to be executed after #3 is complete to pass into the callback
+    uv_queue_work(uv_default_loop(), &baton->request, AsyncShout, (uv_after_work_cb)AfterShout);
+    return;
+}
+
+// this is where we actually exclaim our shout phrase
+void HelloWorld::AsyncShout(uv_work_t* req)
+{
+    AsyncBaton *baton = static_cast<AsyncBaton *>(req->data);
+
+    /***************** custom code here ******************/
+    try
+    {
+
+        std::string return_string = baton->phrase + "!";
+
+        if (baton->louder)
+        {
+            return_string += "!!!!";
+        }
```

Comment:
ah, I see what the issue was @springmeyer - we lost the `baton->result = return_string` here. Adding it back in does the trick.